cpp/benchmarks/streaming/ndsh/q18.cpp

```diff
@@ -0,0 +1,1455 @@
+        if (msg.empty()) {
+            break;
+        }
+        co_await ctx->executor()->schedule();
+        auto chunk = rapidsmpf::ndsh::to_device(
+            ctx, msg.release<rapidsmpf::streaming::TableChunk>()
+        );
+        auto chunk_stream = chunk.stream();
+        auto table = chunk.table_view();
+        total_input_rows += static_cast<std::size_t>(table.num_rows());
+
+        // Semi-join: get indices of matching rows
+        auto match = joiner.semi_join(
+            table.select({key_column_idx}), chunk_stream, ctx->br()->device_mr()
+        );
+
+        // Gather matching rows - convert device_uvector to column_view
+        auto indices = cudf::column_view(
+            cudf::data_type{cudf::type_id::INT32},
+            static_cast<cudf::size_type>(match->size()),
+            match->data(),
+            nullptr,  // null mask
+            0  // null count
+        );
+        auto filtered = cudf::gather(
+            table, indices, cudf::out_of_bounds_policy::DONT_CHECK, chunk_stream
+        );
+
+        total_output_rows += static_cast<std::size_t>(filtered->num_rows());
+
+        if (filtered->num_rows() > 0) {
+            co_await ch_out->send(
+                rapidsmpf::streaming::to_message(
+                    sequence++,
+                    std::make_unique<rapidsmpf::streaming::TableChunk>(
+                        std::move(filtered), chunk_stream
+                    )
+                )
+            );
+        }
+    }
+
+    ctx->comm()->logger().print(
+        "prefilter: rank processed ",
+        total_input_rows,
+        " -> ",
+        total_output_rows,
+        " rows (",
+        (total_output_rows * 100.0 / std::max(total_input_rows, std::size_t{1})),
+        "%)"
+    );
+
+    co_await ch_out->drain(ctx->executor());
+}
+
+/**
+ * @brief Local inner join (both inputs are small after pre-filtering).
+ *
+ * Receives one chunk from each input, joins them, outputs result.
+ */
+rapidsmpf::streaming::Node local_inner_join(
+    std::shared_ptr<rapidsmpf::streaming::Context> ctx,
+    std::shared_ptr<rapidsmpf::streaming::Channel> ch_left,
+    std::shared_ptr<rapidsmpf::streaming::Channel> ch_right,
+    std::shared_ptr<rapidsmpf::streaming::Channel> ch_out,
+    std::vector<cudf::size_type> left_on,
+    std::vector<cudf::size_type> right_on
+) {
+    rapidsmpf::streaming::ShutdownAtExit c{ch_left, ch_right, ch_out};
+
+    auto left_msg = co_await ch_left->receive();
+    auto right_msg = co_await ch_right->receive();
+
+    if (left_msg.empty() || right_msg.empty()) {
+        co_await ch_out->drain(ctx->executor());
+        co_return;
+    }
+
+    auto left_chunk = rapidsmpf::ndsh::to_device(
+        ctx, left_msg.release<rapidsmpf::streaming::TableChunk>()
+    );
+    auto right_chunk = rapidsmpf::ndsh::to_device(
+        ctx, right_msg.release<rapidsmpf::streaming::TableChunk>()
+    );
+
+    auto stream = left_chunk.stream();
```

Comment:
My *guess* is that we need to do something to ensure that `right_table` is valid on `stream` before performing the `cudf::hash_join` on `stream`. We do have `rapidsmpf::cuda_stream_join`, which might work.

(as an aside, I think that this is one of the issues compute-sanitizer is supposed to help with).